### ISSUE TITLE
bug: ion-select firing logging warning/deprecation message to console even when condition not met

### ISSUE THREAD

**Issue by lincolnthree** [ionic-team/ionic-framework#26829]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

- [ ] v4.x
- [ ] v5.x
- [ ] v6.x
- [X] v7.x
- [ ] Nightly

### Current Behavior

ion-select appears to be logging warning messages about using ion-label even when there is no such ion-label element in the DOM. In fact, even a completely unconfigured ion-select component will cause the message to be printed:

![image](https://user-images.githubusercontent.com/362329/220176255-a4d6b8e5-06e5-4b8f-9020-071401ad43e4.png)

### Expected Behavior

Warning & deprecation messages should only be logged when criteria for those messages are met. (Unless you've decided you want to do this for some reason, in which case I defer to your judgement.)

### Steps to Reproduce

<ion-select></ion-select>

### Ionic Info

$ ionic info

Ionic:

   Ionic CLI                     : 6.20.3 (/Users/lincoln/.nvm/versions/node/v16.14.0/lib/node_modules/@ionic/cli)
   Ionic Framework               : @ionic/angular 7.0.0-dev.11676644846.156e8507
   @angular-devkit/build-angular : 15.1.6
   @angular-devkit/schematics    : 15.1.6
   @angular/cli                  : 15.1.6
   @ionic/angular-toolkit        : 8.0.0

Capacitor:

   Capacitor CLI      : 4.6.3
   @capacitor/android : 4.6.3
   @capacitor/core    : 4.6.3
   @capacitor/ios     : 4.6.3

Utility:

   cordova-res : not installed globally
   native-run  : 1.7.1

System:

   NodeJS : v16.14.0 (/Users/lincoln/.nvm/versions/node/v16.14.0/bin/node)
   npm    : 8.3.1
   OS     : macOS Monterey

**Comment by lincolnthree** [comment]
I am also seeing this occur for `ion-range` and `ion-toggle`.

**Comment by liamdebeasi** [comment]
Thanks for the report. This is the intended behavior. Components such as `ion-toggle` or `ion-select` must have an associated label for accessibility reasons. For visible text, developers can use `ion-label`:

```html
<ion-item>
  <ion-label>Toggle Label</ion-label>
  <ion-toggle></ion-toggle>
</ion-item>
```

The new way to do visible text in Ionic 7 is to pass the text into the component itself, removing the need for `ion-label`:

```html
<ion-item>
  <ion-toggle>Toggle Label</ion-toggle>
</ion-item>
```

However, some components may not have visible text. In this case, developers need to use `aria-label` to provide context to screen readers:

```html
<ion-item>
  <ion-toggle aria-label="Toggle Label without visible text"></ion-toggle>
</ion-item>
```

Your code sample uses neither visible text nor an aria label, so we cannot safely assume you are using the modern rendering syntax. Providing either visible text in the component or an `aria-label` should dismiss the warning. Developers who are opted-in to the modern syntax and wish to use the legacy syntax can set `legacy="true"` on the component.

**Comment by lincolnthree** [comment]
Ah thank you. I misread the warning. I got confused by the part about `ion-label` and totally glossed over the case where label is required. Might I propose re-wording this a bit to prevent similar confusion in the future? Or making the warning more specific for the specific deprecation issue that's been encountered?

General rewording:
```
[Ionic Warning]: ion-select now requires either a set [label] attribute, or a set 'aria-label' attribute for compatibility with screen readers. Use of ion-label has been deprecated. To migrate, remove the ion-label and use the appropriate "label" or "aria-label" property on ion-select instead.

Example: <ion-select label="Favorite Color">...</ion-select>

For selects that do not have a visible label, developers should use "aria-label" so screen readers can announce the purpose of the select.

For selects that do not render the label immediately next to the select, developers may continue to use "ion-label" but must manually associate the label with the select by using "aria-labelledby". 
```

**Comment by lincolnthree** [comment]
Additionally, it appears that `[interfaceOptions]={ header: 'My header' }` does not satisfy this constraint. Should it?

**Comment by liamdebeasi** [comment]
No, because `header` is passed to the overlay component (alert, popover) and not set on `ion-select` itself. I can update the deprecation message.

**Comment by lincolnthree** [comment]
Understood. Thank you!

**Comment by liamdebeasi** [comment]
I also added a PR in https://github.com/ionic-team/ionic-framework/pull/26834 to clarify the deprecation messages.

**Comment by lincolnthree** [comment]
Thanks, Liam. I've tried using a few of these options, but both 'aria-label' and 'aria-labelledby' seem to result in a full-width control with no label. (The `ion-toggle`, `ion-segment` etc control takes up the full width of the container, and requires a width being set in order to leave space for the label to grow.)

Is this expected or am I just completely confused?

![Screen Shot 2023-02-21 at 12 18 08 PM](https://user-images.githubusercontent.com/362329/220415099-c32a7bad-9969-47d6-bd84-40a6cba13320.png)


```
  <ion-item>
    <ion-label id="foo">legacy</ion-label>
    <ion-select slot="end" placeholder="foo" [value]="'foo'">
      <ion-select-option value="asdf">asdf</ion-select-option>
    </ion-select>
  </ion-item>
  <ion-item>
    <ion-select placeholder="bar" [value]="'bar'" label="label attr">
      <ion-select-option value="bar">bar</ion-select-option>
    </ion-select>
  </ion-item>
  <ion-item>
    <ion-label id="baz">aria-labelledby attr</ion-label>
    <ion-select slot="end" placeholder="baz" [value]="'baz'" aria-labelledby="baz">
      <ion-select-option value="bar">baz</ion-select-option>
    </ion-select>
  </ion-item>
  <ion-item>
    <ion-select placeholder="cab" [value]="'cab'" aria-label="cab">
      <ion-select-option value="cab">cab</ion-select-option>
    </ion-select>
  </ion-item>
```

Repro:
https://github.com/lincolnthree/ionic-bug-ion-segment/tree/aria-labelledby

**Comment by liamdebeasi** [comment]
`aria-labelledby` should not be used with these components because they use the Shadow DOM. (aria-labelledby does not work across shadow roots). `aria-labelledby` can be used with `ion-input` and `ion-textarea` since they do not use the Shadow DOM.

I'll update our internal logic to not opt-in to the modern syntax in this scenario, but you'll still want to remove the `aria-labelledby` usage here. I also updated the deprecation warning in https://github.com/ionic-team/ionic-framework/pull/26834. Looks like the `aria-labelledby` mention was a copy and paste mistake from `ion-input` and `ion-textarea`.

**Comment by liamdebeasi** [comment]
Can you describe what is broken here?

**Comment by lincolnthree** [comment]
Yeah, in the second 'broken' example, the `ion-select` has grown in height. The only difference was the addition of the `aria-label` attribute.

**Comment by liamdebeasi** [comment]
Ah thanks. We updated the min-heights of `ion-select` in Ionic 7 because it did not match the iOS/MD specs:

https://github.com/ionic-team/ionic-framework/blob/5af0b8c914e18de670b41bcbcbf974e65379d319/core/src/components/select/select.md.scss#L24
https://github.com/ionic-team/ionic-framework/blob/5af0b8c914e18de670b41bcbcbf974e65379d319/core/src/components/select/select.ios.scss#L17

You can account for this by modifying `min-height`.

**Comment by lincolnthree** [comment]
But why should applying the `aria-label` attribute change the height? Isn't that property supposed to be used when an actually visually displayed label does not exist? I'd expect this wouldn't alter the display styling. Am I missing something?

**Comment by liamdebeasi** [comment]
Using `aria-label` opts you in to using the new form syntax. We use some heuristics to do this so developers don't need to manually opt-in in Ionic 7 and then remove that opt-in when Ionic 8 comes around.

So providing a label (either by visible text or an aria-label) will opt you into the new syntax. We provided an escape hatch using `legacy="true"` for when the heuristic gets it wrong.

https://ionicframework.com/docs/v7/api/input#migrating-from-legacy-input-syntax has more information on this.

**Comment by lincolnthree** [comment]
Hmm.

I have to assume using `legacy=true` is not a great long-term solution, though, as I'd expect the legacy escape hatch to be removed in a future version, no?

Thanks for the info. I think I still consider it to be a bug or usability hazard at best -- styles changing when specifying an attribute that is supposed to only impact `a11y` functionality is set seems counter-intuitive. I understand your explanation of 'opting in to new display logic' but I'd expect that only to matter (only be opted in) if you are using a property that impacts the layout like 'label=foo'. 

I'll work on compensating for this in our CSS I guess :)

**Comment by lincolnthree** [comment]
> `aria-labelledby` should not be used with these components because they use the Shadow DOM. (aria-labelledby does not work across shadow roots).

Okay, so following that and we don't use `aria-labelledby`, but rather:

If the `ion-select` component is not contained in an `ion-item`, has no label, and has no `aria-` attributes, an `aria-label` will generated that contains the current value of the `ion-select` -- that's okay -- however, we still get a deprecation message that instructs us to use `aria-label` or `label`.

Now, this becomes problematic because once we do specify an `aria-label` in this scenario, Ionic does something that removes that `aria-label` from the dom entirely. We get no generated attribute **and** our attribute has disappeared.

This is not an issue if `ion-select` is contained within `ion-item`.

**Comment by liamdebeasi** [comment]
> I have to assume using legacy=true is not a great long-term solution, though, as I'd expect the legacy escape hatch to be removed in a future version, no?

That's correct. It is a temporary escape hatch for when a) you are not ready to migrate to the modern syntax and b) our heuristic incorrectly opts you in to the new syntax.

> Thanks for the info. I think I still consider it to be a bug or usability hazard at best -- styles changing when specifying an attribute that is supposed to only impact a11y functionality is set seems counter-intuitive. I understand your explanation of 'opting in to new display logic' but I'd expect that only to matter (only be opted in) if you are using a property that impacts the layout like 'label=foo'.

> IMO. Overriding / extending the a11y properties and using them for layout decisions seems dangerous. It removes (or significantly complicates) developer control of how to design their app while still supporting a11y readers.

This is why the `legacy` escape hatch exists. If we only looked at the `label` prop, then developers who are only using `aria-label` would never be able to opt-in to the modern syntax until Ionic 8. We want to encourage developers to adopt this new syntax before the old syntax is removed. However, we recognize some developers will not be ready to migrate yet.

> I'd expect using either of the aria- attributes would simply squelch the warning messages and meet the requirements for screen readers, nothing more. It sounds like you agree for aria-labelledby, and I'm wondering why not for aria-label.

We removed the `aria-labelledby` requirement because it does not work with Shadow DOM components. So it's not that we want to change the behavior, but it's more that using `aria-labelledby` didn't actually work, it just suppressed the warning message. In other words, the native form control was never associated with the element referenced by `aria-labelledby`.


-----

We realize that the heuristic isn't going to get it right 100% of the time, but the alternative is to have devs manually opt in to the new syntax with `modern="true"` or something similar. We don't like this approach because it essentially penalizes early adopters by a) making them add `modern="true"` and then b) making them remove it in Ionic 8 when the modern syntax is the default.

**Comment by lincolnthree** [comment]
> This is why the `legacy` escape hatch exists.

I totally agree. `modern=true` is not the direction I'd take either. I'm essentially trying to figure out what the end state of this looks like once "migration is complete" and that escape hatch is removed. Trying to figure out it it breaks everything or can be compensated for with CSS.

I haven't found an incantation that works quite yet, but I think I'm getting closer. Thanks for listening, for your patience & time.

**Comment by liamdebeasi** [comment]
Oh another thing to note is the `labelPlacement` and `justify` properties: 

https://ionicframework.com/docs/v7/api/select#label-placement
https://ionicframework.com/docs/v7/api/select#justification

Justify replaces using an `ion-label` with a `slot` in `ion-item`. Label Placement lets you control which order the label/control show up in.

**Comment by lincolnthree** [comment]
Out of curiosity, (and this might help provide some context) do you have a recommended pattern / example for using complex labels like this without using "Legacy" mode? Or is this something that's not really "intended"? Since `aria-labelledby` is not supported for some of these components, what is the alternative? `aria-label`?

![image](https://user-images.githubusercontent.com/362329/220435501-b5de3460-2a90-4431-99c4-50d7056653b9.png)

**Comment by liamdebeasi** [comment]
Something like this should work:

```html
<ion-item>
  <ion-toggle>
    <div>View mode</div>
    <ion-note>Choose how decks are presented visually when viewing or editing.</ion-note>
  </ion-toggle>
</ion-item>
```

Example: https://codepen.io/liamdebeasi/pen/KKxzdEQ

In this case both the "View mode" label header and the sub text ("Choose how decks...") are associated with the control and therefore announced by screen readers.

If you have visible text most of the migration should be moving it from `ion-label` to being passed into the component.

**Comment by liamdebeasi** [comment]
Though for `ion-select` it looks like we only support the `label` property (so plain text only, no HTML). If you created a feature request for passing a label to `ion-select` via slotted content like you can with `ion-toggle` I think it's likely that we would add it.

**Comment by lincolnthree** [comment]
Very interesting. Thank you. This _really is_ a big change. I presume the same would work for `ion-checkbox`?

It looks like `ion-range` already uses the slot syntax. Though... question... would it make sense to support the `slot=label` syntax on ALL of these components, regardless of whether or not they use the shadow dom? That would make things more consistent, even if it's not technically "required". Thoughts?

I'll create a feature request for `ion-select` and link it here.

**Comment by liamdebeasi** [comment]
Yes, the same works for checkbox. Checkbox, Radio, and Select have a default slot to pass in the label. Range only has a named slot (the `'label'` slot) because it has other slots already (`'start'` and `'end'`). Unfortunately, input and textarea do not use the Shadow DOM, so they can't take advantage of slots.

**Comment by lincolnthree** [comment]
Why not send the `default` slot to `label` for all of the input components (for consistency)? 

Also, I didn't think Shadow DOM was required for slotting? I have plenty of components that don't use Shadow DOM that use named slots and/or the default slot. Though this is Angular, not necessarily one of the other frameworks so I don't know if it would work there.

**Comment by liamdebeasi** [comment]
We could, but it increase the boilerplate for certain components. So instead of `<ion-toggle>My Label</ion-toggle>` you'd have to do `<ion-toggle><div slot="label">My Label</div></ion-toggle>`.

Though we have discussed allowing for the `label` property for simple text. If we made them named slots we could add the `label` prop for simpler use cases.


--- 

Web Component slots is a Shadow DOM-only feature. Stencil lets you use slots in scoped components, but it's a bit buggy.

**Comment by lincolnthree** [comment]
RE: Web component slots/shadow dom. Gotcha.

For the rest, I may have been vague. Sorry. I was trying to ask why not send all nested content to the label slot by default, since that seems to be what happens for components that don't use the Shadow DOM?:

```
<ion-toggle>My Label</ion-toggle>
<ion-select>My Label</ion-select>
<ion-select>My Label<div>Also part of the label</div></ion-select>
```
Where all of that gets put into the `label` slot by default.

**Comment by lincolnthree** [comment]
Oh, I guess with `ion-select` you already have `<ion-select-option>` children that need to be put in the right place. So you'd need the `slot=label` designation.

**Comment by lincolnthree** [comment]
Opened an issue to support `slot=label` on `ion-select` instances: https://github.com/ionic-team/ionic-framework/issues/26838

**Comment by liamdebeasi** [comment]
Yeah it's a bit tricky with the slot naming. I think we'd do `ion-select` as a named slot so we could add adornments in the future via slots (i.e. https://github.com/ionic-team/ionic-framework/issues/26297).

Thanks for filing an issue! I'll take a look at that.

**Comment by lincolnthree** [comment]
> I think we'd do `ion-select` as a named slot so we could add adornments in the future via slots (i.e. #26297).

Yeah, this is where I ended up as well. I think everything should just use `slot=label` even if it doesn't strictly require it, technically. More boilerplate? Yes, but also more consistent and more future-proof.

You're welcome! Let me know if I can do anything else to assist. Excited for V7. It solves a lot of problems we've been having with `(ionChange)` events and such, and is overall very nice! **Most** things work after upgrading without any effort at all.
